x/authn/module.go

```diff
@@ -20,18 +27,266 @@ func NewModule() *Module {
 func (m *Module) Initialize(c module.Client) module.Descriptor {
 	m.txDecoder = newTxDecoder()
 
-	return module.NewDescriptorBuilder().
-		Named("authn").
+	builder := module.NewDescriptorBuilder()
+
+	builder.Named("authn").
 		HandlesStateTransition(&v1alpha1.MsgCreateAccount{}, NewCreateAccountController(c), true).
 		HandlesAdmission(&v1alpha1.MsgCreateAccount{}, NewCreateAccountAdmissionController()).
 		OwnsStateObject(&v1alpha1.Account{}, v1alpha1.AccountSchema).
 		OwnsStateObject(&v1alpha1.Params{}, v1alpha1.ParamsSchema).
 		OwnsStateObject(&v1alpha1.CurrentAccountNumber{}, v1alpha1.CurrentAccountNumberSchema).
-		ExtendsAuthentication(extensions2.New(c)).
 		WithGenesis(genesis{c: c}).
-		NeedsStateTransition(&rbacv1alpha1.MsgBindRole{}).Build()
+		NeedsStateTransition(&rbacv1alpha1.MsgBindRole{})
+
+	// add admission controllers, they will only read state
+	// never modify it.
+	builder.
+		WithAuthAdmissionHandler(mempoolFee{}). // verifies if fee matches the minimum
+		WithAuthAdmissionHandler(newAccountExists(c)). // verifies that all signer accounts exist
+		WithAuthAdmissionHandler(newTimeoutBlockExtension(c)). // verifies if tx is not timed-out compared to block
+		WithAuthAdmissionHandler(newValidateMemoExtension(c)). // validates memo length
+		WithAuthAdmissionHandler(newValidateSigCount(c)). // validates number of signatures
+		WithAuthAdmissionHandler(newSigVerifier(c)) // validate signatures
+
+	// add transition controllers for tx, they CAN modify state after
+	// a tx is authenticated
+	builder.
+		WithPostAuthenticationHandler(newConsumeGasForTxSize(c)). // consumes gas for tx size
+		WithPostAuthenticationHandler(newSetPubKeys(c)). // sets pub keys
+		WithPostAuthenticationHandler(newIncreaseSequence(c)) // increases sequence
+
+	return builder.Build()
 }
 
 func (m *Module) GetTxDecoder() authentication.TxDecoder {
 	return m.txDecoder
 }
+
+func newAccountExists(c module.Client) accountExists {
```

Comment:
I'd remove the handlers from module.go, to respect the mantra that module.go can be autogenerated and should contain only what a module does and not the logic of handlers.